lightning/src/routing/router.rs

```diff
@@ -13,29 +13,101 @@
 //! interrogate it to get routes for your own payments.
 
 use bitcoin::secp256k1::PublicKey;
+use bitcoin::hashes::Hash;
+use bitcoin::hashes::sha256::Hash as Sha256;
 
 use crate::ln::channelmanager::ChannelDetails;
 use crate::ln::features::{ChannelFeatures, InvoiceFeatures, NodeFeatures};
 use crate::ln::msgs::{DecodeError, ErrorAction, LightningError, MAX_VALUE_MSAT};
 use crate::routing::gossip::{DirectedChannelInfo, EffectiveCapacity, ReadOnlyNetworkGraph, NetworkGraph, NodeId, RoutingFees};
-use crate::routing::scoring::{ChannelUsage, Score};
+use crate::routing::scoring::{ChannelUsage, LockableScore, Score, ScorerAccountingForInFlightHtlcs};
 use crate::util::ser::{Writeable, Readable, Writer};
 use crate::util::logger::{Level, Logger};
 use crate::util::chacha20::ChaCha20;
 
 use crate::io;
 use crate::prelude::*;
+use crate::sync::Mutex;
 use alloc::collections::BinaryHeap;
 use core::cmp;
 use core::ops::Deref;
 
+/// A [`Router`] implemented using [`find_route`].
+pub struct DefaultRouter<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> where
+	L::Target: Logger,
+	S::Target: for <'a> LockableScore<'a>,
+{
+	network_graph: G,
+	logger: L,
+	random_seed_bytes: Mutex<[u8; 32]>,
+	scorer: S
+}
+
+impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> DefaultRouter<G, L, S> where
+	L::Target: Logger,
+	S::Target: for <'a> LockableScore<'a>,
+{
+	/// Creates a new router using the given [`NetworkGraph`], a [`Logger`], and a randomness source
+	/// `random_seed_bytes`.
+	pub fn new(network_graph: G, logger: L, random_seed_bytes: [u8; 32], scorer: S) -> Self {
+		let random_seed_bytes = Mutex::new(random_seed_bytes);
+		Self { network_graph, logger, random_seed_bytes, scorer }
+	}
+}
+
+impl<G: Deref<Target = NetworkGraph<L>>, L: Deref, S: Deref> Router for DefaultRouter<G, L, S> where
+	L::Target: Logger,
+	S::Target: for <'a> LockableScore<'a>,
+{
+	fn find_route(
+		&self, payer: &PublicKey, params: &RouteParameters, first_hops: Option<&[&ChannelDetails]>,
+		inflight_htlcs: InFlightHtlcs
+	) -> Result<Route, LightningError> {
+		let random_seed_bytes = {
+			let mut locked_random_seed_bytes = self.random_seed_bytes.lock().unwrap();
+			*locked_random_seed_bytes = Sha256::hash(&*locked_random_seed_bytes).into_inner();
+			*locked_random_seed_bytes
+		};
+
+		find_route(
+			payer, params, &self.network_graph, first_hops, &*self.logger,
+			&ScorerAccountingForInFlightHtlcs::new(&mut self.scorer.lock(), inflight_htlcs),
```

Comment:
@jkczyz @TheBlueMatt do we still want to wrap with `ScorerAccountingForInFlightHtlcs` (SAFIFH) here if it's public? If the user-provided `Score` is also a SAFIFH, it will cause us to double-count in-flight htlc usage here

Reply:
I don't think it's a problem because `ScorerAccountingForInFlightHtlcs` takes a `&mut` `Score` whereas `DefaultRouter` takes a `LockableScore`. That is, `ScorerAccountingForInFlightHtlcs` doesn't implement `LockableScore`, so it can't be given to `DefaultRouter`.

Reply:
Ah, that's a good point. I guess it's fine then, I'll update docs + fix the `unreachable`s

Reply:
Oh, right. I guess, though, the user would really need to go out of their way to create one since they would need both a mutable reference to another scorer and an `InFlightHtlcs` object. An empty `InFlightHtlcs` would at least not be harmful if they manage to use `ScorerAccountingForInFlightHtlcs`. Guess they could still shoot themselves in the foot by obtaining a populated one from `ChannelManager`.

I suppose we could simply pass `InFlightHtlcs` into the standalone `find_route` function and use it directly there, and then dispense with `ScorerAccountingForInFlightHtlcs` altogether. That's probably the cleaner approach.

Reply:
Discussed offline, gonna go ahead with a public `ScorerAccountingForInFlightHtlcs` since to jeff's point it's really hard to misuse